# Python Data Wrangling with `pandas`

In this workshop, we provide an introduction to data wrangling with Python. We will do so largely with the `pandas` package, which provides a rich set of tools to manipulate and interact with data frames, the most common data structure used when analyzing tabular data. We'll learn how to manipulate, index, merge, group, and plot data frames using pandas functions.

`pandas` is designed to make it easier to work with structured data. Most of the analyses you might perform will likely involve using tabular data, e.g., from .csv files or relational databases (e.g., SQL). The `DataFrame` object in `pandas` is "a two-dimensional tabular, column-oriented data structure with both row and column labels."

Why's it called `pandas`? The `pandas` website states:

>The `pandas` name itself is derived from *panel data*, an econometrics term for multidimensional structured data sets, and *Python data analysis* itself. After getting introduced, you can consult the full [`pandas` documentation](http://pandas.pydata.org/pandas-docs/stable/).

To motivate this workshop, we'll work with example data and go through the various steps you might need to prepare data for analysis. You'll (hopefully) realize that doing this type of work is much more difficult using Python's built-in data structures.

## The DataFrame: Importing Data and Summary Statistics

To start, we'll work with European unemployment data from Eurostat, which is hosted by [Google](https://code.google.com/p/dspl/downloads/list). There are several `.csv` files that we'll work with in this workshop.

Let's begin by importing `pandas` using the conventional abbreviation.

In [58]:
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
from bs4 import BeautifulStoneSoup
import requests

%matplotlib inline

In [2]:
# Adjust some settings in matplotlib
mpl.rc('savefig', dpi=200)
plt.style.use('ggplot')
plt.rcParams['xtick.minor.size'] = 0
plt.rcParams['ytick.minor.size'] = 0

The `read_csv()` function in `pandas` allows us to easily import our data. By default, it assumes the data is comma-delimited. However, you can specify the delimiter used in your data (e.g., tab, semicolon, pipe, etc.). There are several parameters that you can specify (see the [documentation](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html)). Ultimately, the `read_csv()` function returns a `DataFrame` object, which is the main object `pandas` uses to represent tabular data.

Notice that we call `read_csv()` using the `pd` abbreviation from the import statement above:

In [49]:
unemployment = pd.read_csv('data/country_total.csv')

Great! You've created a `pandas` `DataFrame`. We can look at our data by using the `.head()` method. By default, this shows the header (column names) and the first five rows. Passing an integer, $n$, to `.head()` returns that number of rows. To see the last $n$ rows, use `.tail()`.

In [52]:
unemployment.sample(20)

country seasonality    month  unemployment  unemployment_rate
11561      lt       trend  2009.03        184000               11.2
5737       es       trend  1988.11       2263000               14.9
17811      se          sa  1987.04        105000                2.3
4198       dk          sa  2004.11        155000                5.4
14365      nl       trend  1988.11        397000                6.1
7328       fr          sa  1996.06       2939000               11.5
3334       de          sa  2007.11       3465000                8.0
7861       gr         nsa  2000.02        558000               12.2
9702       ie          sa  1997.10        145000                9.2
16538      pt          sa  2000.06        204000                4.0
3271       de          sa  2002.08       3564000                8.5
9006       hu       trend  1995.10        379000                NaN
9882       ie       trend  1984.10        204000               15.8
3500       de       trend  2002.09       3591000                8.5
878        be         nsa  2002.03        318000                7.3
20788      uk       trend  2010.03       2437000                7.8
1288       be          sa  2008.05        319000                6.7
9644       ie          sa  1992.12        215000               15.7
9842       ie          sa  2009.06        264000               12.1
18827      si       trend  1995.12         68000                NaN

To find the number of rows, you can use the `len()` function. Alternatively, you can use the `shape` attribute.

In [53]:
unemployment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20796 entries, 0 to 20795
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   country            20796 non-null  object 
 1   seasonality        20796 non-null  object 
 2   month              20796 non-null  float64
 3   unemployment       20796 non-null  int64  
 4   unemployment_rate  19851 non-null  float64
dtypes: float64(2), int64(1), object(2)
memory usage: 812.5+ KB


In [6]:
unemployment.shape

(20796, 5)

There are 20,796 rows and 5 columns.

A useful method that generates various summary statistics is `.describe()`.

In [54]:
unemployment.describe()

month  unemployment  unemployment_rate
count  20796.000000  2.079600e+04       19851.000000
mean    1999.401290  7.900818e+05           8.179764
std        7.483751  1.015280e+06           3.922533
min     1983.010000  2.000000e+03           1.100000
25%     1994.090000  1.400000e+05           5.200000
50%     2001.010000  3.100000e+05           7.600000
75%     2006.010000  1.262250e+06          10.000000
max     2010.120000  4.773000e+06          20.900000

You may have noticed that the "count" is lower for the unemployment rate. This is because the summary statistics are based on *non-missing* values. So, this dataset has some missing values for the unemployment rate.

The values depend on what it's called on. If the `DataFrame` includes both numeric and object (e.g., strings) `dtype`s, it will default to summarizing the numeric data. If `.describe()` is called on strings, for example, it will return the count, number of unique values, and the most frequent value along with its count.

---

### Challenge 1: Import Data From A URL

Above, we imported the unemployment data using the `read_csv` function and a relative file path. `read_csv` is [a very flexible method](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.read_csv.html); it also allows us to import data using a URL as the file path. 

A csv file with data on world countries and their abbreviations is located at the URL:

[https://raw.githubusercontent.com/dlab-berkeley/introduction-to-pandas/master/data/countries.csv](https://raw.githubusercontent.com/dlab-berkeley/introduction-to-pandas/master/data/countries.csv)

We've saved this exact URL as a string variable, `countries_url`, below.

Using `read_csv`, import the country data and save it to the variable `countries`.

---

In [64]:
countries_url = 'https://github.com/mhemaly/UDACITY-DECI-Level-03-Data-Science/raw/main/Python-Data-Wrangling/Data/countries.csv'
countries = requests.get(countries_url)
with open("countries.csv", "wb") as file:
    file.write(countries.content)


In [65]:
countries_df = pd.read_csv("countries.csv")

---

### Challenge 2: The `tail` method

DataFrames all have a method called `tail` that takes an integer as an argument and returns a new DataFrame. Before using `tail`, can you guess at what it does? Try using `tail`; was your guess correct?

---

We think it displays the last rows in a DataFrame

In [66]:
# YOUR CODE HERE
countries_df.tail()

country google_country_code country_group         name_en      name_fr  \
25      si                  SI            eu        Slovenia     Slovénie   
26      es                  ES            eu           Spain      Espagne   
27      se                  SE            eu          Sweden        Suède   
28      tr                  TR        non-eu          Turkey      Turquie   
29      uk                  GB            eu  United Kingdom  Royaume-Uni   

                   name_de   latitude  longitude  
25               Slowenien  46.149259  14.986617  
26                 Spanien  39.895013  -2.988296  
27                Schweden  62.198467  14.896307  
28                  Türkei  38.952942  35.439795  
29  Vereinigtes Königreich  54.315447  -2.232612

In [72]:
countries_df

country google_country_code country_group  \
0       at                  AT            eu   
1       be                  BE            eu   
2       bg                  BG            eu   
3       hr                  HR        non-eu   
4       cy                  CY            eu   
5       cz                  CZ            eu   
6       dk                  DK            eu   
7       ee                  EE            eu   
8       fi                  FI            eu   
9       fr                  FR            eu   
10      de                  DE            eu   
11      gr                  GR            eu   
12      hu                  HU            eu   
13      ie                  IE            eu   
14      it                  IT            eu   
15      lv                  LV            eu   
16      lt                  LT            eu   
17      lu                  LU            eu   
18      mt                  MT            eu   
19      nl                  NL            eu   
20      no                  NO        non-eu   
21      pl                  PL            eu   
22      pt                  PT            eu   
23      ro                  RO            eu   
24      sk                  SK            eu   
25      si                  SI            eu   
26      es                  ES            eu   
27      se                  SE            eu   
28      tr                  TR        non-eu   
29      uk                  GB            eu   

                                      name_en  \
0                                     Austria   
1                                     Belgium   
2                                    Bulgaria   
3                                     Croatia   
4                                      Cyprus   
5                              Czech Republic   
6                                     Denmark   
7                                     Estonia   
8                                     Finland   
9                                      France   
10  Germany (including  former GDR from 1991)   
11                                     Greece   
12                                    Hungary   
13                                    Ireland   
14                                      Italy   
15                                     Latvia   
16                                  Lithuania   
17                                 Luxembourg   
18                                      Malta   
19                                Netherlands   
20                                     Norway   
21                                     Poland   
22                                   Portugal   
23                                    Romania   
24                                   Slovakia   
25                                   Slovenia   
26                                      Spain   
27                                     Sweden   
28                                     Turkey   
29                             United Kingdom   

                                              name_fr  \
0                                            Autriche   
1                                            Belgique   
2                                            Bulgarie   
3                                             Croatie   
4                                              Chypre   
5                                  République tchèque   
6                                            Danemark   
7                                             Estonie   
8                                            Finlande   
9                                              France   
10  Allemagne (incluant l'ancienne RDA à partir de...   
11                                              Grèce   
12                                            Hongrie   
13                                            Irlande   
14                                             Italie   
15                                           Lettonie   
16                       

We were correct!

---

### Challenge 3: Describe `countries`

It's important to understand a few fundamentals about your data before you start work with it, including what information it contains, how large it is, and how the values are generally distributed.

Using the methods and attributes above, answer the following questions about `countries`:

* What columns does it contain?
* What does each row stand for?
* How many rows and columns does it contain?
* Are there any missing values in the latitude or longitude columns? 

Hint: the `head` and `describe` functions, as well as the `shape` attribute, will be helpful here.

---

In [69]:
countries_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   country              30 non-null     object 
 1   google_country_code  30 non-null     object 
 2   country_group        30 non-null     object 
 3   name_en              30 non-null     object 
 4   name_fr              30 non-null     object 
 5   name_de              30 non-null     object 
 6   latitude             30 non-null     float64
 7   longitude            30 non-null     float64
dtypes: float64(2), object(6)
memory usage: 2.0+ KB


In [70]:
countries_df.shape

(30, 8)

In [71]:
countries_df.head()

country google_country_code country_group   name_en   name_fr     name_de  \
0      at                  AT            eu   Austria  Autriche  Österreich   
1      be                  BE            eu   Belgium  Belgique     Belgien   
2      bg                  BG            eu  Bulgaria  Bulgarie   Bulgarien   
3      hr                  HR        non-eu   Croatia   Croatie    Kroatien   
4      cy                  CY            eu    Cyprus    Chypre      Zypern   

    latitude  longitude  
0  47.696554  13.345980  
1  50.501045   4.476674  
2  42.725674  25.482322  
3  44.746643  15.340844  
4  35.129141  33.428682

#### Answers

1. country, google_country_code, ... (see .info())
2. It stands for a country
3. 30 rows , 8 columns
4. No.

## Rename, Index, and Slice

In the unemployment dataset, you may have noticed that the `month` column also includes the year. Let's go ahead and rename it:

In [73]:
unemployment.rename(columns={'month' : 'year_month'}, inplace=True)

The `.rename()` method allows you to modify index labels and/or column names. As you can see, we passed a `dict` to the `columns` parameter, with the original name as the key and the new name as the value. Importantly, we also set the `inplace` parameter to `True`, which modifies the *actual* `DataFrame`, not a copy of it.

It might also make sense to separate the data in `year_month` into two separate columns. To do this, we'll need to access a single column. There are two ways to do this: we can either use bracket (`[]`) or dot notation (referred to as *attribute access*).

In [74]:
# Bracket notation
unemployment['year_month'].head()

0    1993.01
1    1993.02
2    1993.03
3    1993.04
4    1993.05
Name: year_month, dtype: float64

In [75]:
# Dot notation
unemployment.year_month.head()

0    1993.01
1    1993.02
2    1993.03
3    1993.04
4    1993.05
Name: year_month, dtype: float64

It is preferrable to use the bracket notation as a column name might inadvertently have the same name as a `DataFrame` (or `Series`) method. In addition, only bracket notation can be used to create a new column. If you try and use attribute access to create a new column, you'll create a new attribute, *not* a new column.

When selecting a single column, we have a `Series` object, which is a single vector of data (e.g., a NumPy array) with "an associated array of data labels, called its *index*." A `DataFrame` also has an index. In our example, the indices are an array of sequential integers, which is the default. You can find them in the left-most position, without a column label.

Indices need not be a sequence of integers. They can, for example, be dates or strings. Note that indices do *not* need to be unique.

Indices, like column names, can be used to select data. However, indices can be used to select particular rows (as opposed to columns). In fact, you can do something like `.head()` with slicing using the `[]` operator.

In [79]:
unemployment[:5]

country seasonality  year_month  unemployment  unemployment_rate
0      at         nsa     1993.01        171000                4.5
1      at         nsa     1993.02        175000                4.6
2      at         nsa     1993.03        166000                4.4
3      at         nsa     1993.04        157000                4.1
4      at         nsa     1993.05        147000                3.9

Before we continue, let's look at a few useful ways to index data&mdash;that is, select rows.

`.loc` primarily works with string labels. It accepts a single label, a list (or array) of labels, or a slice of labels (e.g., `'a' : 'f'`).

Let's create a `DataFrame` to see how this works. 

In [77]:
# Create a bacteria dataframe
bacteria = pd.DataFrame(
    {'bacteria_counts' : [632, 1638, 569, 115],
    'other_feature' : [438, 833, 234, 298]},
    index=['Firmicutes',
           'Proteobacteria',
           'Actinobacteria',
           'Bacteroidetes'])

Notice that we pass in a `dict`, where the keys correspond to column names and the values to the data. In this example, we've also set the indices&mdash;strings in this case&mdash;to be the taxon of each bacterium.

In [80]:
bacteria

bacteria_counts  other_feature
Firmicutes                  632            438
Proteobacteria             1638            833
Actinobacteria              569            234
Bacteroidetes               115            298

Now, if we're interested in the values (row) associated with "Actinobacteria," we can use `.loc` and the index name.

In [81]:
bacteria.loc['Actinobacteria']

bacteria_counts    569
other_feature      234
Name: Actinobacteria, dtype: int64

This returns the column values for the specified row. Interestingly, we could have also used "positional indexing," even though the indices are strings.

In [82]:
bacteria[2:3]

bacteria_counts  other_feature
Actinobacteria              569            234

The difference is that the former returns a `Series` because we selected a single lable, while the latter returns a `DataFrame` because we selected a range of positions.

Let's return to our unemployment data. Another indexing option, `.iloc`, primarily works with integer positions. To select specific rows, we can do the following.

In [83]:
unemployment.iloc[[1, 5, 6, 9]]

country seasonality  year_month  unemployment  unemployment_rate
1      at         nsa     1993.02        175000                4.6
5      at         nsa     1993.06        134000                3.5
6      at         nsa     1993.07        128000                3.4
9      at         nsa     1993.10        141000                3.7

We can select a range of rows and specify the step value.

In [84]:
unemployment.iloc[25:50:5]

country seasonality  year_month  unemployment  unemployment_rate
25      at         nsa     1995.02        174000                4.5
30      at         nsa     1995.07        123000                3.3
35      at         nsa     1995.12        175000                4.7
40      at         nsa     1996.05        159000                4.3
45      at         nsa     1996.10        146000                3.9

(Note: As is typical in Python, the end position is not included. Therefore, we don't see the row associated with the index 50.)

Indexing is important. You'll use it a lot. Below, we'll show how to index based on data values.



---

### Challenge 4: Renaming a Column

The "other_feature" column in our `bacteria` table isn't very descriptive. Suppose we know that "other_feature" refers to a second set of bacteria count observations. Use the `rename` method to give "other_feature" a more descriptive name.

---

In [87]:
# YOUR CODE HERE
bacteria.rename(columns={"other_feature": "second_bacteria_counts"},inplace = True)

---

### Challenge 5: Indexing to Obtain a Specific Value

Both `loc` and `iloc` can be used to select a particular value if they are given two arguments. The first argument is the name (when using `loc`) or index number (when using `iloc`) of the *row* you want, while the second argument is the name or index number of the *column* you want.

Using `loc`, select "Bacteroidetes" and "bacteria_counts" to get the count of Bacteroidetes.

BONUS: how could you do the same task using `iloc`?

---

In [92]:
# YOUR CODE HERE
bacteria.loc['Bacteroidetes'].bacteria_counts

115

In [93]:
bacteria.iloc[3].bacteria_counts

115

---

### Challenge 6: Indexing Multiple Rows and Columns

Both `loc` and `iloc` can be used to select subsets of columns *and* rows at the same time if they are given lists (and/or slices, for `iloc`] as their two arguments. 

Using `iloc` on the `unemployment` DataFrame, get:
* every row starting at row 4 and ending at row 7
* the 0th, 2nd, and 3rd columns

BONUS: how could you do the same task using `loc`?

---

In [96]:
# YOUR CODE HERE
unemployment.iloc[4:8]

country seasonality  year_month  unemployment  unemployment_rate
4      at         nsa     1993.05        147000                3.9
5      at         nsa     1993.06        134000                3.5
6      at         nsa     1993.07        128000                3.4
7      at         nsa     1993.08        130000                3.4

In [106]:
unemployment.iloc[4:8,[0,2,3]]

country  year_month  unemployment
4      at     1993.05        147000
5      at     1993.06        134000
6      at     1993.07        128000
7      at     1993.08        130000

In [102]:
unemployment.loc[4:7]

country seasonality  year_month  unemployment  unemployment_rate
4      at         nsa     1993.05        147000                3.9
5      at         nsa     1993.06        134000                3.5
6      at         nsa     1993.07        128000                3.4
7      at         nsa     1993.08        130000                3.4

In [108]:
unemployment.loc[4:8,['country','year_month','unemployment']]

country  year_month  unemployment
4      at     1993.05        147000
5      at     1993.06        134000
6      at     1993.07        128000
7      at     1993.08        130000
8      at     1993.09        132000

## Manipulating Columns: Renaming, Creating, Reordering

So, we still want to split `year_month` into two separate columns. Above, we saw that this column is type (technically, `dtype`) `float64`. We'll first extract the year using the `.astype()` method. This allows for type casting&mdash;basically converting from one type to another. We'll then subtract this value from `year_month`&mdash;to get the decimal portion of the value&mdash;and multiply the result by 100 and convert to `int`.

For more information on `pandas` `dtype`s, check the documentation [here](http://pandas.pydata.org/pandas-docs/stable/basics.html#dtypes).

In [109]:
unemployment['year'] = unemployment['year_month'].astype(int)

In this case, we're casting the floating point values to integers. In Python, this [truncates the decimals](https://docs.python.org/2/library/stdtypes.html#numeric-types-int-float-long-complex).

Finally, let's create our **month** variable as described above. (Because of the truncating that occurs when casting to `int`, we first round the values to the nearest whole number.)

In [110]:
unemployment['month'] = ((unemployment['year_month'] - unemployment['year']) * 100).round(0).astype(int)

In [111]:
unemployment.head(12)

country seasonality  year_month  unemployment  unemployment_rate  year  \
0       at         nsa     1993.01        171000                4.5  1993   
1       at         nsa     1993.02        175000                4.6  1993   
2       at         nsa     1993.03        166000                4.4  1993   
3       at         nsa     1993.04        157000                4.1  1993   
4       at         nsa     1993.05        147000                3.9  1993   
5       at         nsa     1993.06        134000                3.5  1993   
6       at         nsa     1993.07        128000                3.4  1993   
7       at         nsa     1993.08        130000                3.4  1993   
8       at         nsa     1993.09        132000                3.5  1993   
9       at         nsa     1993.10        141000                3.7  1993   
10      at         nsa     1993.11        156000                4.1  1993   
11      at         nsa     1993.12        169000                4.4  1993   

    month  
0       1  
1       2  
2       3  
3       4  
4       5  
5       6  
6       7  
7       8  
8       9  
9      10  
10     11  
11     12

To create the `month` column, we subtracted two vectors. This resulted in the decimal value in `year_month`. To transform the values to integers, we multiplied by 100.

Now, let's say we wanted to **reorder the columns** in the `DataFrame`. For this, we use bracket notation again, passing in a list of column names in the order we'd like to see them.

In [112]:
unemployment = unemployment[['country',
                             'seasonality',
                             'year_month',
                             'year',
                             'month',
                             'unemployment',
                             'unemployment_rate']]

In [113]:
unemployment.head(10)

country seasonality  year_month  year  month  unemployment  \
0      at         nsa     1993.01  1993      1        171000   
1      at         nsa     1993.02  1993      2        175000   
2      at         nsa     1993.03  1993      3        166000   
3      at         nsa     1993.04  1993      4        157000   
4      at         nsa     1993.05  1993      5        147000   
5      at         nsa     1993.06  1993      6        134000   
6      at         nsa     1993.07  1993      7        128000   
7      at         nsa     1993.08  1993      8        130000   
8      at         nsa     1993.09  1993      9        132000   
9      at         nsa     1993.10  1993     10        141000   

   unemployment_rate  
0                4.5  
1                4.6  
2                4.4  
3                4.1  
4                3.9  
5                3.5  
6                3.4  
7                3.4  
8                3.5  
9                3.7

---

### Challenge 7: Another Way to Obtain the Year

If you didn't know that casting floats to ints truncates the decimals in Python, you could have used NumPy's `floor()` function. `np.floor` takes an array or `pd.Series` of floats as its argument, and returns an array or `pd.Series` where every float has been rounded down to the nearest whole number. 

Use `np.floor` to round the values in the `year_month` column down so we can cast them as integer years. Note that the types are still floats, so we'll still need to use `astype` to typecast.

---

In [116]:
import numpy as np

# Select the "year_month" column
year_month = unemployment['year_month']

# Use np.floor on year_month to get the years as floats
years_by_floor = np.floor(year_month)

# Cast years_by_floor to integers using astype(int)
int_years = years_by_floor.astype(int)

# Check that this gets the same answers as our first approach
# This should return True
(unemployment['year_month'].astype(int) == int_years).all()

True

The last line of code in the previous cell does an element-wise comparison of the values in the corresponding arrays. The `.all()` method checks whether *all* elements are `True`.

## Merging DataFrames

So far, our `DataFrame` is organized in a reasonable way. But, we know we can do better. We're eventually going to be interested in the unemployment rate for each country. The trouble is, we don't exactly know what the values in `country` refer to. We can fix that by getting country names. These can be found in `countries.csv`. Run the following cell to import the countries data using a URL as the file path.

In [118]:
# countries_url = 'https://github.com/mhemaly/UDACITY-DECI-Level-03-Data-Science/blob/main/Python-Data-Wrangling/Data/countries.csv'
countries = pd.read_csv("countries.csv")

In [119]:
countries.tail(3)

country google_country_code country_group         name_en      name_fr  \
27      se                  SE            eu          Sweden        Suède   
28      tr                  TR        non-eu          Turkey      Turquie   
29      uk                  GB            eu  United Kingdom  Royaume-Uni   

                   name_de   latitude  longitude  
27                Schweden  62.198467  14.896307  
28                  Türkei  38.952942  35.439795  
29  Vereinigtes Königreich  54.315447  -2.232612

This file has lots of useful information. It even has the country names is three different languages.

Because the data we need is stored in two separate files, we'll want to merge the data somehow. Let's determine which column we can use to join this data. `country` looks like a good option. However, we don't need all of the columns in the `countries` `DataFrame`. To select certain columns, we use the name bracket notation we used to reorder the columns.

In [120]:
country_names = countries[['country', 'country_group', 'name_en']]

In [121]:
country_names.head(2)

country country_group  name_en
0      at            eu  Austria
1      be            eu  Belgium

`pandas` includes an easy-to-use merge function. Let's use it to merge the two `DataFrame`s on country code.

In [122]:
unemployment = pd.merge(unemployment, country_names, on='country')

Merging is often more complex than this example. If you want to merge on multiple columns, you can pass a list of column names to the `on` parameter.

```
pd.merge(first, second, on=['name', 'id'])
```

For more information on merging, check the [documentation](http://pandas.pydata.org/pandas-docs/stable/merging.html#database-style-dataframe-joining-merging).

`pandas` also provides a `.merge()` method that can act on a `DataFrame`. You can read more about that [here](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html).

---

### Challenge 8

You may sometimes need to merge on columns with different names. To do so, use the `left_on` and `right_on` parameters, where the first listed `DataFrame` is the "left" one and the second is the "right." It might look something this:

```
pd.merge(one, two, left_on='city', right_on='city_name')
```

Suppose wanted to merge `unemployment` with a new DataFrame called `country_codes`, where the abbreviation for each country is in the column "c_code":

---

In [123]:
country_codes = country_names.rename({"country": "c_code"}, axis=1).drop("country_group", axis=1)
country_codes.head()

c_code   name_en
0     at   Austria
1     be   Belgium
2     bg  Bulgaria
3     hr   Croatia
4     cy    Cyprus

Use `merge` to merge `unemployment` and `country_codes` on their country codes. Make sure to specify `left_on=` and `right_on=` in the call to `merge`!

In [125]:
unemployment = pd.merge(unemployment, country_codes, left_on='country',right_on="c_code")

## Calculating Unique and Missing Values

That's better. We now know that the abbreviation "at" corresponds to Austria. We might be curious to check what countries we have data for. The `Series` object includes a `.unique()` method. We'll use this to check the countries. We can select the name either using bracket or dot notation. (While we suggested using brackets above, it *is* sometimes easier to use dot notation. Just be careful.)

In [126]:
unemployment.name_en.unique()

array(['Austria', 'Belgium', 'Bulgaria', 'Cyprus', 'Czech Republic',
       'Germany (including  former GDR from 1991)', 'Denmark', 'Estonia',
       'Spain', 'Finland', 'France', 'Greece', 'Croatia', 'Hungary',
       'Ireland', 'Italy', 'Lithuania', 'Luxembourg', 'Latvia', 'Malta',
       'Netherlands', 'Norway', 'Poland', 'Portugal', 'Romania', 'Sweden',
       'Slovenia', 'Slovakia', 'Turkey', 'United Kingdom'], dtype=object)

To get a count of the **number of unique countries,** we can either wrap the above code with `len()` to get the number of items in the array, or we can use the  `Series.nunique()` method.

In [ ]:
unemployment.name_en.nunique()

It might be more interesting to know how many observations we actually have per unique value. `pandas` has a `Series` method called `.value_counts()` that returns the counts for the unique values in the `Series`.

In [34]:
unemployment['name_en'].value_counts()

KeyError: 'name_en'

By default, the `Series` is sorted by values. If you'd like it sorted by index&mdash;country name in this case&mdash;append the `.sort_index()` method.

In [ ]:
unemployment['name_en'].value_counts().sort_index()

This will be useful for our analysis. The maximum number of observations for a given country for this time period is 1,008 observations. We'll note that certain countries, such as Turkey, have far less data.

How about finding the **date range** for this data set? Let's look at the minimum and maximum years.

In [ ]:
unemployment['year'].min(), unemployment['year'].max()

Next, we should pause for a moment and think about what data we really care about. For our purposes, the variable of interest is `unemployment_rate`. The number of observations by country only reflect the number of instances of each country name in the dataset. It is possible, maybe even expected, to have some missing data within those instances. Let's find out **how many unemployment rate values are missing.**

In [ ]:
unemployment['unemployment_rate'].isnull().sum()

The `.isnull()` method returns a corresponding boolean value for each entry in the unemployment rate `Series`. In Python `True` is equivalent to 1 and `False` is equivalent to 0. Thus, when we add the result (with `.sum()`), we get a count for the *total* number of missing values.

---

### Challenge 9: Exploring Unemployment Rates

What are the minimum and maximum unemployment rates in our data set? Which unemployment rates are most and least common?

Hint: look at where we found the minimum and maximum years for a hint to the first question, and use `value_counts` for the second.

---

In [35]:
# YOUR CODE HERE

## Aggregating Statistics Across Groups: Group By

What if we'd like to know how many missing values exist at the *country* level? To do so, we need to calculate a separate value for each unique value of country. In other words, we're going to **group** our data **by** a specific column, and calculate some quantity within each group. The "group-by" operation is a fundamental technique used with relational dataset.

First, we can take the main part of what we had above and create a new column in the `DataFrame`:

In [36]:
unemployment['unemployment_rate_null'] = unemployment['unemployment_rate'].isnull()

To count the number of missing values by each country, we'll perform the grouping operation with the `.groupby()` method:

In [37]:
unemployment.groupby('name_en')['unemployment_rate_null'].sum()

KeyError: 'name_en'

Let's explain what just happened. We start with our `DataFrame`. We tell `pandas` that we want to group the data by country name&mdash;that's what goes in the parentheses. The output of the `groupby()` function is a special object in `pandas` called a `GroupBy` object.

The `GroupBy` object doesn't have what we want, because we need to tell the object how we want to **aggregate**. In this case, we're going to aggregate by summing the missing values across each group -- i.e., the aggregation is the `sum()` operation. So, we need to tell it what column we'd like to perform the `.sum()` operation on. In this case, it's the indicator for whether or not the unemployment rate was missing.

As we saw above, the number of records for each country differs. We might, then, want to have the missing values by country shown as proportions. Let's do a separate `groupby()` calculation. This time, we'll use a the `mean()` function as the aggregator, since we want to to calculate proportions for each country:

In [ ]:
unemployment.groupby('name_en')['unemployment_rate_null'].mean()

This looks good. Let's save this as its own `DataFrame`. We'll re-run the `groupby()`, but pass in a new argument called `as_index=False`. What does this argument do?

In [ ]:
unemployment_rate = unemployment.groupby('name_en', as_index=False)['unemployment_rate_null'].mean()

In [38]:
unemployment_rate.head()

NameError: name 'unemployment_rate' is not defined

`unemployment_rate` is a `DataFrame` with the null proportion for unemployment rates. It's important to note that using `as_index=False` in `.groupby()` only works if the grouping column(s) are not the same as the columns on which we're performing the operation.

Also, to group by several columns, simply pass in a list of column names to `.groupby()`.

```
unemployment.groupby(['name_en', 'seasonality'])['unemployment_rate'].mean()
```

---

### Challenge 10: Group By Practice

Find the average unemployment rate for European Union vs. non-European Union countries. 

1. First, use `groupby()` to group on "country_group".
2. Then, select the "unemployment_rate" column,
3. Aggregate by using `.mean()` to get the average.

---

In [39]:
# YOUR CODE HERE

## Exporting A DataFrame to a CSV File

From our analysis on missing unemployment rates, we can observe that Croatia is missing a relatively large amount of data (~66% of records are missing). The `unemployment_rate` data frame contains this information, and is useful to keep for further analysis.

Suppose we wanted to save this as a .csv file. For this, we'd use the `.to_csv()` method:

In [40]:
unemployment_rate.to_csv('../data/unemployment_missing.csv')

NameError: name 'unemployment_rate' is not defined

Let's import this newly created file, and see what it looks like:

In [ ]:
unemployment_missing = pd.read_csv('../data/unemployment_missing.csv')
unemployment_missing.head()

What do you see here? By default, this method writes the indices. We probably don't want that. Let's edit the code. Let's also be explicit about the type of delimiter we're interested in. (Values can be separated by pipes (`|`), semicolons (`;`), tabs (`\t`), etc.)

In [ ]:
unemployment_rate.to_csv('../data/unemployment_missing.csv', index=False, sep=',')

In [41]:
unemployment_missing = pd.read_csv('../data/unemployment_missing.csv')
unemployment_missing.head()

FileNotFoundError: [Errno 2] No such file or directory: '../data/unemployment_missing.csv'

Much better!

Let's return to our main `DataFrame`. Now that we have the missing values information in `unemployment_rate`, we can **drop the last column** we added to `unemployment`.

In [ ]:
unemployment.drop('unemployment_rate_null', axis=1, inplace=True)

It's important to specify the `axis` parameter. Specifically, `axis=1` refers to columns while `axis=0` refers to rows. The parameter `inplace=True` simply modifies the actual `DataFrame` rather than returning a new `DataFrame`.

In [42]:
unemployment.head()

country seasonality  year_month  year  month  unemployment  \
0      at         nsa     1993.01  1993      1        171000   
1      at         nsa     1993.02  1993      2        175000   
2      at         nsa     1993.03  1993      3        166000   
3      at         nsa     1993.04  1993      4        157000   
4      at         nsa     1993.05  1993      5        147000   

   unemployment_rate  unemployment_rate_null  
0                4.5                   False  
1                4.6                   False  
2                4.4                   False  
3                4.1                   False  
4                3.9                   False

## Dealing With Missing Values: Boolean Indexing

Now that we know about the missing values, we have to deal with them. There are two main options:

* Fill the missing values with some other values.
* Do not use observations with missing values.
    * Depending on the analysis, we may want to exclude entire countries.
    
Because countries with missing unemployment rate data have at least 36 missing values, which is too many to fill, we'll take the second approach and **exclude missing values** from our primary analyses.

Instead of just getting rid of that data, it might make sense to store it in a separate `DataFrame`. This way, we could answer questions such as, "do missing values occur during certain months (or years) more frequently?" With this, we will introduce the concept of *boolean indexing* for filtering data.

Recall that we used the `isnull()` function to determine whether a data element is missing. This function returns a `pd.Series` containing Boolean values:

In [43]:
unemployment['unemployment_rate'].isnull()

0        False
1        False
2        False
3        False
4        False
         ...  
20791    False
20792    False
20793    False
20794    False
20795    False
Name: unemployment_rate, Length: 20796, dtype: bool

To subselect the rows with missing values, we'll use this Boolean `pd.Series` to index the `unemployment` data frame, just like we might row numbers or column names. Specifically, we'll pass the Boolean series into a pair of bracket, which will then return only the rows where the value in the array is `True`. Let's see what this looks like:

In [44]:
unemployment_rate_missing = unemployment[unemployment['unemployment_rate'].isnull()]

In [45]:
unemployment_rate_missing.head()

country seasonality  year_month  year  month  unemployment  \
1656      bg         nsa     1995.01  1995      1        391000   
1657      bg         nsa     1995.02  1995      2        387000   
1658      bg         nsa     1995.03  1995      3        378000   
1659      bg         nsa     1995.04  1995      4        365000   
1660      bg         nsa     1995.05  1995      5        346000   

      unemployment_rate  unemployment_rate_null  
1656                NaN                    True  
1657                NaN                    True  
1658                NaN                    True  
1659                NaN                    True  
1660                NaN                    True

It is also possible to specify multiple conditions using the `&` operator, but each condition needs to be inside of parentheses. The `.isin()` method, which takes a `list` of values, is useful when you're interested in conditioning on multiple values on a given column. For example, if you want to select multiple countries.

Now, we're ready to remove the missing data in `unemployment`. To do this, we can use the `.dropna()` method.

In [46]:
unemployment.dropna(subset=['unemployment_rate'], inplace=True)

Note that while we have dropped all observations for which `unemployment_rate == NaN`, this doesn't mean that all of our observations overlap exactly in time. We may find that there are dates where we have data for one country and no data for others.

---

### Challenge 11: Boolean Indexing

Suppose we only want to look at unemployment data from the year 2000 or later. Use Boolean indexing to create a DataFrame with only these years.

1. Select the "year" column from `unemployment`.
2. Using the year data, create a **mask**: an array of Booleans where each value is True if and only if the year is 2000 or later. Remember, you can use Boolean operators like `>`, `<`, and `==` on a column.
3. Use the mask from step 2 to index `unemployment`.

---

In [47]:
# Select the year column from unemployment
year = # YOUR CODE HERE

# Create a mask
later_or_equal_2000 = # YOUR CODE HERE

# Boolean index unemployment
# YOUR CODE HERE

SyntaxError: invalid syntax (2703933329.py, line 2)

## Sorting Values

At this point, you might be curious to know what the highest unemployment rates were. For this, we'll use the `sort_values()` method to sort the data.

In [ ]:
unemployment.sort_values('unemployment_rate', ascending=False)[:5]

The above code creates a copy of the `DataFrame`, sorted in *descending* order, and prints the first five rows.

You may have noticed that the data set includes a `seasonality` column, which we haven't yet discussed. The unemployment rate in this data is actually calculated in three separate ways. Let's look at the values.

In [48]:
unemployment['seasonality'].unique()

array(['nsa', 'sa', 'trend'], dtype=object)

The three options above correspond to:

* nsa: not seasonally adjusted
* sa: seasonally adjusted
* trend: trend cycle

We'll stick with seasonally adjusted data so that the values are more comparable. To do this, we'll create a Boolean index. Then, we'll sort the values by unemployment. We'll set `ascending=False` so that it's in descending order, and we see the highest unemployment rates on top:

In [ ]:
unemployment[unemployment['seasonality'] == 'sa'].sort_values('unemployment_rate', ascending=False)[:5]

Spain has the highest seasonally adjusted unemployment rate.



Thanks & Best Regards </p>
[Eng \ Mahmoud Elhemaly](https://www.linkedin.com/in/mahmoudelhemaly/overlay/contact-info/)